File: src/main/java/me/miki/shindo/gui/mainmenu/MainMenuScene.java
package me.miki.shindo.gui.mainmenu;

import lombok.Getter;
import me.miki.shindo.Shindo;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.color.ColorManager;
import me.miki.shindo.management.color.palette.ColorPalette;
import me.miki.shindo.management.color.palette.ColorType;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import net.minecraft.client.Minecraft;

import java.awt.Color;

public class MainMenuScene {

    private static final ColorPalette FALLBACK_PALETTE = new ColorPalette();
    private static final AccentColor FALLBACK_ACCENT = new AccentColor(
            "Default",
            new Color(170, 255, 169),
            new Color(17, 255, 189)
    );

    @Getter
    private final GuiShindoMainMenu parent;
    @Getter
    private final SimpleAnimation animation = new SimpleAnimation();

    public Minecraft mc = Minecraft.getMinecraft();

    public MainMenuScene(GuiShindoMainMenu parent) {
        this.parent = parent;
    }

    public void initScene() {
    }

    public void initGui() {
    }

    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
    }

    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
    }

    public void keyTyped(char typedChar, int keyCode) {
    }

    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
    }

    public void handleInput() {
    }

    public void onGuiClosed() {
    }

    public void onSceneClosed() {
    }

    public void setCurrentScene(MainMenuScene scene) {
        parent.setCurrentScene(scene);
    }

    public Color getBackgroundColor() {
        return getMenuPalette().getBackgroundColor(ColorType.DARK);
    }

    protected Color getPanelColor() {
        return getMenuPalette().getBackgroundColor(ColorType.MID);
    }

    protected Color getControlColor() {
        return getMenuPalette().getBackgroundColor(ColorType.NORMAL);
    }

    protected ColorPalette getMenuPalette() {
        ColorManager manager = getColorManager();
        return manager != null ? manager.getPalette() : FALLBACK_PALETTE;
    }

    protected AccentColor getMenuAccent() {
        ColorManager manager = getColorManager();
        AccentColor accent = manager != null ? manager.getCurrentColor() : null;
        return accent != null ? accent : FALLBACK_ACCENT;
    }

    private ColorManager getColorManager() {
        Shindo instance = Shindo.getInstance();
        return instance != null ? instance.getColorManager() : null;
    }

    public MainMenuScene getSceneByClass(Class<? extends MainMenuScene> clazz) {
        return parent.getSceneByClass(clazz);
    }
}

File: src/main/kotlin/me/miki/shindo/gui/mainmenu/MainMenuScene.kt
package me.miki.shindo.gui.mainmenu

import me.miki.shindo.Shindo
import me.miki.shindo.management.color.AccentColor
import me.miki.shindo.management.color.ColorManager
import me.miki.shindo.management.color.palette.ColorPalette
import me.miki.shindo.management.color.palette.ColorType
import me.miki.shindo.ui.animation.v2.value.SimpleAnimation
import net.minecraft.client.Minecraft
import java.awt.Color

open class MainMenuScene(
    private val parent: GuiShindoMainMenu,
) {
    val animation = SimpleAnimation()

    val mc: Minecraft = Minecraft.getMinecraft()

    open fun initScene() {
    }

    open fun initGui() {
    }

    open fun drawScreen(
        mouseX: Int,
        mouseY: Int,
        partialTicks: Float,
    ) {
    }

    open fun mouseClicked(
        mouseX: Int,
        mouseY: Int,
        mouseButton: Int,
    ) {
    }

    open fun keyTyped(
        typedChar: Char,
        keyCode: Int,
    ) {
    }

    open fun mouseReleased(
        mouseX: Int,
        mouseY: Int,
        mouseButton: Int,
    ) {
    }

    open fun handleInput() {
    }

    open fun onGuiClosed() {
    }

    open fun onSceneClosed() {
    }

    fun setCurrentScene(scene: MainMenuScene?) {
        parent.setCurrentScene(scene)
    }

    fun getBackgroundColor(): Color = getMenuPalette().getBackgroundColor(ColorType.DARK)

    protected fun getPanelColor(): Color = getMenuPalette().getBackgroundColor(ColorType.MID)

    protected fun getControlColor(): Color = getMenuPalette().getBackgroundColor(ColorType.NORMAL)

    protected fun getMenuPalette(): ColorPalette {
        val manager = getColorManager()
        return manager.getPalette()
    }

    protected fun getMenuAccent(): AccentColor {
        val manager = getColorManager()
        return manager.getCurrentColor()
    }

    private fun getColorManager(): ColorManager {
        val instance = Shindo.getInstance()
        return instance.getColorManager()
    }

    fun getSceneByClass(clazz: Class<out MainMenuScene>): MainMenuScene? = parent.getSceneByClass(clazz)

    fun getParent(): GuiShindoMainMenu = parent

    companion object {
        private val FALLBACK_PALETTE = ColorPalette()
        private val FALLBACK_ACCENT =
            AccentColor(
                "Default",
                Color(170, 255, 169),
                Color(17, 255, 189),
            )
    }
}
